Add first section to `CONTRIBUTING.md` (#147)
SharezoneApp/sharezone-app: CONTRIBUTING.md

# Contributing to Sharezone
_Note: The `CONTRIBUTING.md` is in the process of making. See [#28](https://github.com/SharezoneApp/sharezone-app/issues/28) for the missing sections._

## Welcome
We love, that you are interested in contributing to Sharezone 💙 There are many ways to contribute and we appreciate all of them. This document gives a rough overview of how you can contribute to Sharezone and which steps you need to follow to set up the development environment.

* [How to set up your development environment](#how-to-set-up-your-development-environment)

If you have ans questions, please join our [Discord](https://sharezone.net/discord).

## How to set up your development environment
If you would like to contribute 

### Operating system
_tl:dr: We recommend macOS._

You can use the operating system you like. But we recommend to use macOS because you might have some issues with other operating systems.

> **Known issues:**

Author: @Jonas-Sander Are we supporting (supporting = we know it's working for the platform and willing to fix issues) all platforms for the Sharezone CLI (like macOS, ChromeOS, etc.)? Or do we only support macOS and (maybe) Windows? If we only supported specific operating systems, we should add this to our known issues, like:

> Sharezone CLI is only supported for macOS. Although, the CLI should work for other operating system but might have some issues. See here (a link to all GitHub tickets for specifc label for the CLI) for all known issues with the CLI for other operating systems.

Reviewer: The CLI should support Windows and macOS and I think it does right now.

Author: I added the following point to `Known issues`:

> Sharezone CLI has official support for macOS and Windows (should also work with other operating systems, but might cause problems in some cases)

> * Golden tests are only passing with macOS

Author: At the moment, we don't have any golden tests. But #138 introduce the first golden tests of our repository. I will add later an issue which describes why this issue exists and what the workarounds are for other operating systems. After creating the issue, I will add a link to the workarounds to this point.

We added some files of the `.vscode` folder to this repository. This ensures that all VS Code users have the same VS Code settings (like formatting, launch configurations, etc.) for this repository.

### Flutter
Our client uses the [Flutter](https://flutter.dev) framework. In order to run tests, compile the app, etc. you need to setup Flutter.

Please follow the official documentation: [Install Flutter](https://docs.flutter.dev/get-started/install)

### Sharezone CLI
We written our own CLI to manage our repository. Common use cases for the CLI are:
* Get all Flutter/Dart packages for all packages inside this repository (`sz packages get`)
* Run all tests for all packages inside this repository (`sz test`)
* Analyze all packages inside this repository (`sz analyze`)

Execute the following steps to install the Sharezone CLI:
1. [Clone this repository](#clone-this-repository)
2. Navigate to the repository (`cd sharezone-app`)
3. Run `pub global activate --source path "tools/sz_repo_cli/"`

> 4. Add the `./bin` to your environment variables

Author: @Jonas-Sander Are these steps correct?

Reviewer: For macOS yes, but the files in `bin` are only usable for macOS right now. For windows one would need to invoke the script manually i think. I guess there should be some way to make it work for windows but idk how (maybe adding a `sz.bat` and `sharezone.bat`? not sure though)

Author: Like Flutter does it for the Flutter CLI? https://github.com/flutter/flutter/blob/master/bin/flutter.bat

Reviewer: I guess, yeah

Author: Ok. For now, Windows users should use `dart run tools/sz_repo_cli/bin/sz_repo_cli.dart`. I split the installation steps for the Sharezone CLI into Windows and macOS in [b2c53c6](https://github.com/SharezoneApp/sharezone-app/pull/147/commits/b2c53c694a7a45b24610cb02fceac973f201220a).

# Contributing to Sharezone
_Note: The `CONTRIBUTING.md` is in the process of making. See [#28](https://github.com/SharezoneApp/sharezone-app/issues/28) for the missing sections._

## Welcome
We love, that you are interested in contributing to Sharezone 💙 There are many ways to contribute and we appreciate all of them. This document gives a rough overview of how you can contribute to Sharezone and which steps you need to follow to set up the development environment.

* [How to set up your development environment](#how-to-set-up-your-development-environment)

> If you have ans questions, please join our [Discord](https://sharezone.net/discord).

Reviewer: ```suggestion
If you have any questions, please join our [Discord](https://sharezone.net/discord).
```

## How to set up your development environment
If you would like to contribute to our code, please follow the introduction to set up your development environment.

### Operating system
_tl:dr: We recommend macOS._

You can use the operating system you like. But we recommend to use macOS because you might have some issues with other operating systems.

**Known issues:**
* Golden tests are only passing with macOS

> * Sharezone CLI has official support for macOS and Windows (should also work with other operating systems, but might cause problems in some cases)

Reviewer: ```suggestion
* The Sharezone CLI (used for development) only officially supports macOS and Windows (should also work with other operating systems, but might cause problems in some cases)
```

Just so it makes more sense as a "known issue". Also added the "(used for development)" since the general explanation for the CLI is located further below and hasn't been read yet.

If you discover unknown issues which are related to the operating system, please submit a new ticket on [GitHub](https://github.com/SharezoneApp/sharezone-app/issues/new/choose).

### IDE
_tl:dr: We recommend VS Code._

Flutter supports IDE support for [VS Code](https://code.visualstudio.com/) and [Android Studio](https://developer.android.com/studio). We support only VS Code. This doesn't mean you can't use other IDEs like Android Studio. But you might need set up more configuration steps which aren't documented by us (like launch configurations).

We added some files of the `.vscode` folder to this repository. This ensures that all VS Code users have the same VS Code settings (like formatting, launch configurations, etc.) for this repository.

### Flutter

> Our client uses the [Flutter](https://flutter.dev) framework. In order to run tests, compile the app, etc. you need to setup Flutter.

Reviewer: Nit:
```suggestion
Our Sharezone app uses the [Flutter](https://flutter.dev) framework. In order to run tests, compile the app, etc. you need to setup Flutter.
```

I mean client is also correct but I think this is just easier to understand since we always speak of the "app" and not really of our "Sharezone client".

Please follow the official documentation: [Install Flutter](https://docs.flutter.dev/get-started/install)

### Sharezone CLI
We written our own CLI to manage our repository. Common use cases for the CLI are:
* Get all Flutter/Dart packages for all packages inside this repository (`sz packages get`)
* Run all tests for all packages inside this repository (`sz test`)
* Analyze all packages inside this repository (`sz analyze`)

Execute the following steps to install the Sharezone CLI:
1. [Clone this repository](#clone-this-repository)
2. Navigate to the repository (`cd sharezone-app`)

> 3. Run `pub global activate --source path "tools/sz_repo_cli/"`

Reviewer: I don't even think you need to run `pub global` as the `sz`/`sharezone` files in `/bin` will run `dart run [...]/sz_repo_cli.dart` directly.

Author: Yes, this makes sense. I just copied I from our CI. What was the reason to run `pub global activate --source path ...` in our CI?

https://github.com/SharezoneApp/sharezone-app/blob/f84ff9ddb1b76d9b81c114c056199091ea11f5a8/.github/workflows/main.yml#L42

Reviewer: Good question :D  
I think it is a remain from our Gitlab Pipeline which doesn't make sense anymore.  
I'm gonna try the pipeline without it locally, gimme a minute.

Reviewer: Ah havent tried it yet but maybe so it automatically gets the dependencies?

Author: So running `dart pub get -C tools/sz_repo_cli` instead of `pub global activate --source path "$CI_CD_DART_SCRIPTS_PACKAGE_PATH"` works?

Reviewer: Yes

Author: Ok, then I will use `dart pub get -C tools/sz_repo_cli` because less confusing 👍

### Sharezone CLI
We written our own CLI to manage our repository. Common use cases for the CLI are:
* Get all Flutter/Dart packages for all packages inside this repository (`sz packages get`)
* Run all tests for all packages inside this repository (`sz test`)
* Analyze all packages inside this repository (`sz analyze`)

**macOS**
Execute the following steps to install the Sharezone CLI:
1. [Clone this repository](#clone-this-repository)
2. Navigate to the repository (`cd sharezone-app`)
3. Run `dart pub get -C tools/sz_repo_cli`
4. Add the `./bin` to your environment variables
5. Restart your terminal

You should now be able to run `sz` or `sharezone` in your terminal. 

**Windows**
At the moment, there is no Windows support for a command alias like `sz` or `sharezone`. Instead you need to run `dart run tools/sz_repo_cli/bin/sz_repo_cli.dart`, like `dart run tools/sz_repo_cli/bin/sz_repo_cli.dart packages get`. Please keep this in mind when reading commands like `sz packages get`.

Execute the following steps to install the Sharezone CLI:
1. [Clone this repository](#clone-this-repository)
2. Navigate to the repository (`cd sharezone-app`)

> You should now be able to run `dart run tools/sz_repo_cli/bin/sz_repo_cli.dart`.

Reviewer: Shouldn't there also be pub get here?
